File: bindings/ruby/lib/strling/native_client.rb
# frozen_string_literal: true

require 'fiddle'
require 'json'
require 'pathname'
require 'thread'

module Strling
  # Reentrant native client over caller-selected strling.c-abi v1.
  class NativeClient
    STATUS_RESPONSE_WRITTEN = 0
    private_constant :STATUS_RESPONSE_WRITTEN

    def self.load(library_path)
      path = Pathname.new(String(library_path))
      raise NativeAdapterError.new(:native_load, 'native STRling library path must be absolute') unless path.absolute?

      normalized = path.cleanpath
      unless normalized.file?
        raise NativeAdapterError.new(:native_load, "native STRling library not found: #{normalized}")
      end

      new(normalized.to_s)
    rescue NativeAdapterError
      raise
    rescue StandardError => e
      raise NativeAdapterError.new(:native_load, "invalid native STRling library path: #{e.message}")
    end

    attr_reader :library_path

    def initialize(library_path)
      @library_path = library_path.freeze
      @mutex = Mutex.new
      @condition = ConditionVariable.new
      @active_calls = 0
      @closed = false
      load_functions
    end

    def closed?
      @mutex.synchronize { @closed }
    end

    def execute(request)
      encoded = JSON.generate(request, allow_nan: false).encode(Encoding::UTF_8)
      raise NativeAdapterError.new(:transport, 'interop request is not valid UTF-8') unless encoded.valid_encoding?
      if encoded.bytesize > MAX_INTEROP_REQUEST_BYTES
        raise NativeAdapterError.new(:transport, "interop request exceeds #{MAX_INTEROP_REQUEST_BYTES} bytes")
      end

      begin_call
      begin
        execute_bytes(encoded)
      ensure
        end_call
      end
    rescue JSON::GeneratorError, EncodingError, TypeError => e
      raise NativeAdapterError.new(:transport, "interop request is not strict JSON: #{e.message}")
    end

    def describe
      completed_result(envelope('describe', {}))
    end

    def compile(compile_request, target_profile: nil)
      payload = { 'compile_request' => compile_request }
      payload['target_profile'] = target_profile unless target_profile.nil?
      completed_result(envelope('compile', payload))
    end

    def inspect_target_profile(target_profile)
      completed_result(envelope('target_profile.inspect', 'target_profile' => target_profile))
    end

    def simply_compile(builder_request, target_profile: nil)
      payload = { 'builder_request' => builder_request }
      payload['target_profile'] = target_profile unless target_profile.nil?
      completed_result(envelope('simply.compile', payload))
    end

    def close
      handle = @mutex.synchronize do
        return nil if @closed

        @closed = true
        @condition.wait(@mutex) until @active_calls.zero?
        selected = @handle
        @handle = nil
        selected
      end
      handle.close if handle&.respond_to?(:close)
      nil
    end

    private

    def load_functions
      # Fiddle does not expose RTLD_LOCAL on every supported Ruby even though
      # local visibility is the native loader default when RTLD_GLOBAL is absent.
      @handle = Fiddle::Handle.new(@library_path, Fiddle::RTLD_NOW)
      @abi = function('strling_interop_abi_version_v1', [], Fiddle::TYPE_UINT)
      @execute = function(
        'strling_interop_execute_v1',
        [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T, Fiddle::TYPE_VOIDP],
        Fiddle::TYPE_UINT
      )
      @free = function(
        'strling_interop_owned_bytes_free_v1',
        [Fiddle::TYPE_VOIDP],
        Fiddle::TYPE_UINT
      )
      actual = @abi.call
      return if actual == NATIVE_ABI_VERSION

      @handle.close
      @handle = nil
      raise NativeAdapterError.new(:native_abi, "expected strling.c-abi #{NATIVE_ABI_VERSION}", status: actual)
    rescue Fiddle::DLError => e
      @handle&.close
      @handle = nil
      raise NativeAdapterError.new(:native_load, "cannot load native STRling library: #{e.message}")
    end

    def function(symbol, arguments, result)
      Fiddle::Function.new(@handle[symbol], arguments, result)
    end

    def begin_call
      @mutex.synchronize do
        raise NativeAdapterError.new(:closed_client, 'native STRling client is closed') if @closed

        @active_calls += 1
      end
    end

    def end_call
      @mutex.synchronize do
        @active_calls -= 1
        @condition.broadcast if @active_calls.zero?
      end
    end

    def execute_bytes(encoded)
      input = Fiddle::Pointer.malloc(encoded.bytesize, Fiddle::RUBY_FREE)
      input[0, encoded.bytesize] = encoded unless encoded.empty?
      descriptor_size = Fiddle::SIZEOF_VOIDP + Fiddle::SIZEOF_SIZE_T
      output = Fiddle::Pointer.malloc(descriptor_size, Fiddle::RUBY_FREE)
      output[0, descriptor_size] = "\0" * descriptor_size
      primary_error = nil
      begin
        status = @execute.call(encoded.empty? ? 0 : input, encoded.bytesize, output)
        unless status == STATUS_RESPONSE_WRITTEN
          raise NativeAdapterError.new(:native_abi, 'native STRling execution failed', status: status)
        end
        address = output[0, Fiddle::SIZEOF_VOIDP].unpack1(pointer_pack)
        length = output[Fiddle::SIZEOF_VOIDP, Fiddle::SIZEOF_SIZE_T].unpack1(size_pack)
        if address.zero? || length.zero? || length > MAX_INTEROP_RESPONSE_BYTES
          raise NativeAdapterError.new(:transport, 'native STRling returned an invalid or oversized response')
        end
        raw = Fiddle::Pointer.new(address)[0, length]
        decode_response(raw)
      rescue StandardError => e
        primary_error = e
        raise
      ensure
        release_status = @free.call(output)
        if primary_error.nil? && release_status != STATUS_RESPONSE_WRITTEN
          raise NativeAdapterError.new(:native_abi, 'native STRling response release failed', status: release_status)
        end
      end
    end

    def pointer_pack
      Fiddle::SIZEOF_VOIDP == 8 ? 'Q' : 'L'
    end

    def size_pack
      Fiddle::SIZEOF_SIZE_T == 8 ? 'Q' : 'L'
    end

    class StrictObject < Hash
      def []=(key, value)
        raise JSON::ParserError, "duplicate property #{key.inspect}" if key?(key)

        super
      end
    end
    private_constant :StrictObject

    def decode_response(raw)
      text = raw.dup.force_encoding(Encoding::UTF_8)
      raise NativeAdapterError.new(:transport, 'native STRling response is not strict UTF-8') unless text.valid_encoding?

      value = JSON.parse(text, object_class: StrictObject, create_additions: false)
      unless value.is_a?(Hash) && value['interop_protocol_version'] == INTEROP_PROTOCOL_VERSION
        raise NativeAdapterError.new(:transport, 'interop response has an unsupported version')
      end
      status = value['status']
      unless %w[completed error].include?(status)
        raise NativeAdapterError.new(:transport, 'interop response has an unsupported status')
      end
      if status == 'completed'
        raise NativeAdapterError.new(:transport, 'completed interop response has no result') unless value.key?('result')
      else
        error = value['error']
        unless error.is_a?(Hash) && error['code'].is_a?(String) && error['path'].is_a?(String)
          raise NativeAdapterError.new(:transport, 'failed interop response has no stable code and path')
        end
      end
      value
    rescue JSON::ParserError => e
      raise NativeAdapterError.new(:transport, "native STRling response is not strict JSON: #{e.message}")
    end

    def envelope(operation, payload)
      {
        'interop_protocol_version' => INTEROP_PROTOCOL_VERSION,
        'operation' => operation,
        'payload' => payload
      }
    end

    def completed_result(request)
      response = execute(request)
      raise InteropProtocolError, response if response['status'] == 'error'

      response['result']
    end
  end
end

File: bindings/swift/Sources/STRling/NativeClient.swift
import CSTRlingNative
import Foundation

public let interopProtocolVersion = "1.0.0"
public let nativeABIVersion: UInt32 = 1
public let maxInteropRequestBytes = 10_485_760
public let maxInteropResponseBytes = 33_554_432

public enum NativeErrorKind: String, Sendable {
    case load
    case abi
    case closed
    case transport
}

public struct NativeAdapterError: Error, CustomStringConvertible, Sendable {
    public let kind: NativeErrorKind
    public let detail: String
    public let status: UInt32?

    public init(kind: NativeErrorKind, detail: String, status: UInt32? = nil) {
        self.kind = kind
        self.detail = detail
        self.status = status
    }

    public var description: String {
        if let status { return "\(kind.rawValue): \(detail) (status \(status))" }
        return "\(kind.rawValue): \(detail)"
    }
}

public struct InteropProtocolError: Error, CustomStringConvertible, Sendable {
    public let code: String
    public let path: String
    public let operation: String?

    public var description: String { "\(code) at \(path)" }
}

public final class NativeClient: @unchecked Sendable {
    public let libraryPath: String

    private let lifecycle = NSCondition()
    private var handle: UnsafeMutableRawPointer?
    private var activeCalls = 0
    private var closed = false

    public init(libraryPath: String) throws {
        guard !libraryPath.isEmpty,
              URL(fileURLWithPath: libraryPath).path == libraryPath,
              FileManager.default.fileExists(atPath: libraryPath)
        else {
            throw NativeAdapterError(
                kind: .load,
                detail: "native STRling library path must be an existing absolute file"
            )
        }
        self.libraryPath = libraryPath
        var error = [CChar](repeating: 0, count: 512)
        let opened = libraryPath.withCString { pointer in
            error.withUnsafeMutableBufferPointer { buffer in
                strling_swift_native_open(pointer, buffer.baseAddress, buffer.count)
            }
        }
        guard let opened else {
            let detail = error.withUnsafeBufferPointer { buffer in
                String(cString: buffer.baseAddress!)
            }
            throw NativeAdapterError(kind: .load, detail: detail)
        }
        let actual = strling_swift_native_abi_version(opened)
        guard actual == nativeABIVersion else {
            strling_swift_native_close(opened)
            throw NativeAdapterError(
                kind: .abi,
                detail: "expected strling.c-abi \(nativeABIVersion)",
                status: actual
            )
        }
        handle = opened
    }

    deinit { close() }

    public var isClosed: Bool {
        lifecycle.lock()
        defer { lifecycle.unlock() }
        return closed
    }

    public func execute(_ request: [String: Any]) throws -> [String: Any] {
        guard JSONSerialization.isValidJSONObject(request) else {
            throw NativeAdapterError(kind: .transport, detail: "interop request is not strict JSON")
        }
        let encoded = try JSONSerialization.data(withJSONObject: request, options: [.sortedKeys])
        guard encoded.count <= maxInteropRequestBytes else {
            throw NativeAdapterError(
                kind: .transport,
                detail: "interop request exceeds \(maxInteropRequestBytes) bytes"
            )
        }
        let current = try beginCall()
        defer { endCall() }

        var output = strling_swift_owned_bytes(data: nil, len: 0)
        let result: Result<[String: Any], Error>
        do {
            let status = encoded.withUnsafeBytes { raw in
                strling_swift_native_execute(
                    current,
                    raw.bindMemory(to: UInt8.self).baseAddress,
                    encoded.count,
                    &output
                )
            }
            guard status == 0 else {
                throw NativeAdapterError(
                    kind: .abi,
                    detail: "native STRling execution failed",
                    status: status
                )
            }
            guard let data = output.data,
                  output.len > 0,
                  output.len <= maxInteropResponseBytes
            else {
                throw NativeAdapterError(
                    kind: .transport,
                    detail: "native STRling returned an invalid or oversized response"
                )
            }
            let copied = Data(bytes: data, count: output.len)
            result = .success(try Self.decodeResponse(copied))
        } catch {
            result = .failure(error)
        }
        let freeStatus = strling_swift_native_free(current, &output)
        if freeStatus != 0, case .success = result {
            throw NativeAdapterError(
                kind: .abi,
                detail: "native STRling response release failed",
                status: freeStatus
            )
        }
        return try result.get()
    }

    public func describe() throws -> Any {
        try completedResult(envelope("describe", [:]))
    }

    public func compile(
        _ compileRequest: [String: Any],
        targetProfile: [String: Any]? = nil
    ) throws -> Any {
        var payload: [String: Any] = ["compile_request": compileRequest]
        if let targetProfile { payload["target_profile"] = targetProfile }
        return try completedResult(envelope("compile", payload))
    }

    public func inspectTargetProfile(_ targetProfile: [String: Any]) throws -> Any {
        try completedResult(
            envelope("target_profile.inspect", ["target_profile": targetProfile])
        )
    }

    public func simplyCompile(
        _ builderRequest: [String: Any],
        targetProfile: [String: Any]? = nil
    ) throws -> Any {
        var payload: [String: Any] = ["builder_request": builderRequest]
        if let targetProfile { payload["target_profile"] = targetProfile }
        return try completedResult(envelope("simply.compile", payload))
    }

    public func close() {
        lifecycle.lock()
        if closed {
            lifecycle.unlock()
            return
        }
        closed = true
        while activeCalls != 0 { lifecycle.wait() }
        let current = handle
        handle = nil
        lifecycle.unlock()
        strling_swift_native_close(current)
    }

    private func beginCall() throws -> UnsafeMutableRawPointer {
        lifecycle.lock()
        defer { lifecycle.unlock() }
        guard !closed, let handle else {
            throw NativeAdapterError(kind: .closed, detail: "native STRling client is closed")
        }
        activeCalls += 1
        return handle
    }

    private func endCall() {
        lifecycle.lock()
        activeCalls -= 1
        if activeCalls == 0 { lifecycle.broadcast() }
        lifecycle.unlock()
    }

    private func completedResult(_ request: [String: Any]) throws -> Any {
        let response = try execute(request)
        if response["status"] as? String == "error",
           let body = response["error"] as? [String: Any],
           let code = body["code"] as? String,
           let path = body["path"] as? String
        {
            throw InteropProtocolError(
                code: code,
                path: path,
                operation: response["operation"] as? String
            )
        }
        guard let result = response["result"] else {
            throw NativeAdapterError(
                kind: .transport,
                detail: "completed interop response has no result"
            )
        }
        return result
    }

    private func envelope(_ operation: String, _ payload: [String: Any]) -> [String: Any] {
        [
            "interop_protocol_version": interopProtocolVersion,
            "operation": operation,
            "payload": payload,
        ]
    }

    private static func decodeResponse(_ data: Data) throws -> [String: Any] {
        let value: Any
        do {
            guard String(data: data, encoding: .utf8) != nil else {
                throw NativeAdapterError(
                    kind: .transport,
                    detail: "native STRling response is not UTF-8"
                )
            }
            value = try JSONSerialization.jsonObject(with: data, options: [])
            var validator = JSONDuplicateKeyValidator(data)
            try validator.validate()
        } catch {
            throw NativeAdapterError(
                kind: .transport,
                detail: "native STRling response is not strict UTF-8 JSON: \(error)"
            )
        }
        guard let response = value as? [String: Any],
              response["interop_protocol_version"] as? String == interopProtocolVersion,
              let status = response["status"] as? String,
              status == "completed" || status == "error"
        else {
            throw NativeAdapterError(
                kind: .transport,
                detail: "interop response has an unsupported version or status"
            )
        }
        if status == "completed", response["result"] == nil {
            throw NativeAdapterError(
                kind: .transport,
                detail: "completed interop response has no result"
            )
        }
        if status == "error" {
            guard let body = response["error"] as? [String: Any],
                  body["code"] is String,
                  body["path"] is String
            else {
                throw NativeAdapterError(
                    kind: .transport,
                    detail: "failed interop response has no stable code and path"
                )
            }
        }
        return response
    }
}

private struct JSONDuplicateKeyValidator {
    private let bytes: [UInt8]
    private var index = 0

    init(_ data: Data) { bytes = Array(data) }

    mutating func validate() throws {
        skipWhitespace()
        try value()
        skipWhitespace()
        guard index == bytes.count else { throw ValidationError.trailingContent }
    }

    private mutating func value() throws {
        skipWhitespace()
        guard index < bytes.count else { throw ValidationError.missingValue }
        switch bytes[index] {
        case 0x7b: try object()
        case 0x5b: try array()
        case 0x22: _ = try string()
        default:
            while index < bytes.count,
                  ![0x2c, 0x5d, 0x7d, 0x20, 0x09, 0x0a, 0x0d].contains(bytes[index])
            {
                index += 1
            }
        }
    }

    private mutating func object() throws {
        index += 1
        skipWhitespace()
        var names = Set<String>()
        if take(0x7d) { return }
        while true {
            skipWhitespace()
            let name = try string()
            guard names.insert(name).inserted else {
                throw ValidationError.duplicateProperty(name)
            }
            skipWhitespace()
            try expect(0x3a)
            try value()
            skipWhitespace()
            if take(0x7d) { return }
            try expect(0x2c)
        }
    }

    private mutating func array() throws {
        index += 1
        skipWhitespace()
        if take(0x5d) { return }
        while true {
            try value()
            skipWhitespace()
            if take(0x5d) { return }
            try expect(0x2c)
        }
    }

    private mutating func string() throws -> String {
        let start = index
        try expect(0x22)
        var escaped = false
        while index < bytes.count {
            let character = bytes[index]
            index += 1
            if escaped {
                escaped = false
            } else if character == 0x5c {
                escaped = true
            } else if character == 0x22 {
                return try JSONDecoder().decode(String.self, from: Data(bytes[start..<index]))
            }
        }
        throw ValidationError.unterminatedString
    }

    private mutating func skipWhitespace() {
        while index < bytes.count, [0x20, 0x09, 0x0a, 0x0d].contains(bytes[index]) {
            index += 1
        }
    }

    private mutating func take(_ expected: UInt8) -> Bool {
        guard index < bytes.count, bytes[index] == expected else { return false }
        index += 1
        return true
    }

    private mutating func expect(_ expected: UInt8) throws {
        guard take(expected) else { throw ValidationError.unexpectedToken(expected) }
    }

    private enum ValidationError: Error {
        case duplicateProperty(String)
        case missingValue
        case trailingContent
        case unexpectedToken(UInt8)
        case unterminatedString
    }
}
